WordPress/gutenberg
# Fix issue where resetBlocks can result in an incorrect block selection

## Description
Fixes #21445. Alternative to #21371. Bug discovered while working on #21340.

When RESET_BLOCKS is dispatched and results in the selected blocks being replaced, selection state in the store is not cleared, which can cause the top toolbar to be shown for a block that no longer exists.

<strike>This PR converts the resetBlocks action into a generator action, and adds additional logic for resetting selection when needed.</strike>

This PR adds a higher order reducer for selection that updates the selection state as necessary when RESET_BLOCKS is dispatched.

## How has this been tested?
Unit tests added.

Steps to reproduce:
1. Enable the Top Toolbar option
2. Add a paragraph with some text to a post and ensure it stays selected so that the block toolbar is shown at the top
3. In the browser dev tools open up the console and execute the following to reset blocks: 
```js
wp.data.dispatch('core/block-editor').resetBlocks([]);
```
4. Observe in the top toolbar the block toolbar is no longer displayed. Previously the movers and more menu were displayed and clicking the more menu caused the editor to crash.

## Types of changes

Bug fix (non-breaking change which fixes an issue)



## Checklist:
